perly - simplify non-terminal bare_statement_for

happy-barney opened the pull request:
By extracting various declarations of cursor variable(s) into dedicated non-terminal
for statement now contains only 3 branches:
- `for (;;)`
- `for ()`
- `for cursor ()`

with LALR(1) parser `cursor` cannot be empty -> it will produce conflict with `for (;;)` form.

* This set of changes does not require a perldelta entry.

Commit:
[parser] Add `clause_mexpr` to remove duplicated code

Sequence `PERLY_PAREN_OPEN` `mexpr` `PERLY_PAREN_CLOSE` is repeated
in several places, add new non-terminal `clause_mexpr` to remove
this repetition.

Commit:
[parser] Add `clause_mexpr` to remove duplicated code

Sequence `PERLY_PAREN_OPEN` `mexpr` `PERLY_PAREN_CLOSE` is repeated
in several places, add new non-terminal `clause_mexpr` to remove
this repetition.

Commit:
[parser] Rename `for-itervars` variable(s) to `itervars`

To visualize duplicated code.

Commit:
[parser] Rename `for-itervars` variable(s) to `itervars`

To visualize duplicated code.

bram-perl commented on perly.y:
```diff
@@ -346,8 +346,8 @@ bare_statement_field_declaration
 
 bare_statement_for
 	:	KW_FOR
-		PERLY_PAREN_OPEN
 		remember
+		PERLY_PAREN_OPEN
```
**Commit [parser] bare_statement_for - move remember right after keyword forming sequence `KW_FOR remember`.**

This commit message is still a bit too cryptic for me.

> Non-terminal `remember` is empty non-terminal with action starting
> new full lexical scope. Such empty non-terminals, when used in
> multiple branches, will generate shift/reduce conflict when not
> appearing after distinguishable sequence of terminals.

I simply can't understand most of that..
Reading the `Such empty non-terminals, when used in multiple branches, will generate shift/reduce conflict` makes me wonder if there is/was an issue with how it was written before and/or if there is any change to how the construct is now treated.
 
 
> Change formalizes `KW_FOR` as a keyword starting new lexical scope.

The commit message should also what the consequence of this is. (Does it change anything? If yes then what? If no then explicitly mention it in the commit message).

Also missing a bit of the "why"; What was the real problem (if any) with the old code?

bram-perl commented on perly.y:
```diff
@@ -119,6 +119,7 @@
 %type <opval> bare_statement_when
 %type <opval> bare_statement_while
 %type <opval> bare_statement_yadayada
+%type <opval> clause_mexpr
```
**Commit [parser] clause_mexpr - deduplicate shared pattern into non-terminal PERLY_PAREN_OPEN mexpr PERLY_PAREN_CLOSE**

For this commit I would use something like:
```
Add `clause_mexpr` to remove duplicated code

The pattern `PERLY_PAREN_OPEN mexpr `PERLY_PAREN_CLOSE` is repeated in several places
add `clause_mexpr` to avoid this repetition.

This change should not affect any of the behaviour. 
```

In my opinion having: `[parser] clause_mexpr` and `non-terminal` and `PERLY_PAREN_OPEN mexpr PERLY_PAREN_CLOSE` in the commit title does not add any real value.  (The title is too long and reading it in `git log --oneline` I would really need to think hard what it really means and if it might alter behavior or not)

bram-perl commented on perly.y:
```diff
@@ -380,39 +380,44 @@ bare_statement_for
 		remember
 		KW_MY
-		my_scalar
+		my_scalar[cursor]
```
**Commit [parser] bare_statement_for - rename variable(s) product to cursor to visualize possible deduplication**

To me (but I'm unfamiliar with the parser) `variable(s) product` is meaningless/confusing.

I also find the word `cursor` confusing (but again could be because I'm unfamiliar with the code and/or non-native English). I don't know what to expect when I see the word `cursor`..

tonycoz replied:
It made me think of a cursor in a database sense, which lets you iterate over the results of a database query, just as foreach style loops iterate over a list.

bram-perl commented on perly.y:
```diff
@@ -119,6 +119,7 @@
 %type <opval> bare_statement_when
 %type <opval> bare_statement_while
 %type <opval> bare_statement_yadayada
+%type <opval> clause_for_cursor
```
**Commit [parser] clause_for_cursor - encapsulate FOR loop cursor variables**

This is already an improvement;
What I would still add in the commit message is what branches are now merged.

I.e. something like:
```
Merge the for-branches of:
- `for $scalar (...)`
- `for my $scalar (...)`
- `for my ($foo, $bar) (...)`
- `for \ $scalar (...)`
- `for my \ $scalar (..)`

into a single branch with the parsing of the variables handled  by `clause_for_cusror`
```

Commit:
[tests] New assert function `assume` with test message first

* Motivation

Introduce single assert function which can:
- use emulated named arguments
- evaluate code to build "got" value
- detect eval success/failure
- detect expected eval success/failure

`assume` function enforces most important part of test case,
assumed behaviour description (test message), first.

Commit:
[tests] New assert function `assume` with test message first

* Motivation

Introduce single assert function which can:
- use emulated named arguments
- evaluate code to build "got" value
- detect eval success/failure
- detect expected eval success/failure

`assume` function enforces most important part of test case,
assumed behaviour description (test message), first.

Commit:
[tests] New tests focused on scope behaviour of loop variables

be it control variable of `for (;;)` variant or cursor variable(s)
of `for $var () { } continue { }`

Tests test:
- visibility of variable in for loop
- visibility of variable in continue loop
- behaviour after loop

Commit:
[tests] New tests focused on scope behaviour of loop variables

be it control variable of `for (;;)` variant or cursor variable(s)
of `for $var () { } continue { }`

Tests test:
- visibility of variable in for loop
- visibility of variable in continue loop
- behaviour after loop

Commit:
[parser] Split `for` statement into two variants

for_cstyle - C-style with control expression
for_itervars - iterating over list using variable(s)

Commit:
[parser] Split `for` statement into two variants

for_cstyle - C-style with control expression
for_itervars - iterating over list using variable(s)

happy-barney commented:
@bram-perl I've updated PR and commit messages, can you please look whether it is better described ?

bram-perl commented on t/test.pl:
```diff
@@ -52,6 +57,49 @@ sub _print_stderr {
     print STDERR @_;
 }
 
+sub assume {
```
**Commit [tests] New assert function assume with test message first**

This needs more documentation but not sure where/what the proper place is. \
At the very least I expect to see comments to indicate what can keys are supported in `%args`.

A quick look at the places where it is used: personally I'm not a big fan of it :/

happy-barney replied:
Comments added.

I already extend this function in my next branch with keys:
- `fresh_perl => {}` - when provided, uses `fresh_perl` instead of `eval` (I'm trying to fix segfault)
- `todo => q (...)` - when provided, given `assert` is marked as todo
- `experiment => q (name)` - when set, enables given experiment for evaluated code

Regarding usage:

Function's name comes for BDD approach.

Main reason I'm using (even have module like this: Test::YAFT) it is getting rid of boring boilerplate "lives_ok / dies_ok / value assert" polluting tests with unnecessary complexity, when single `expect` / `throws` is enough. 

For example, following pattern:
```
TODO: {
    local $::TODO = $todo_message;
    todo_skip $skip_reason if $skip_condition;
    fresh_perl ($code, $args);
   is (0, $?, q (executed ok));
   like ($results, qr (regex));
}
```

can be written like:
```
assume q (...)
    => todo => $todo_message
    => skip  => sub { $skip_reason if $skip_condition }
    => fresh_perl => $args
    => expect => qr (regex)
    => eval => $code
    ;
```

and once todo functionality is implemented, you can simply remove just `todo` and maybe `fresh_perl` lines.

bram-perl commented on MANIFEST:
```diff
@@ -6351,7 +6351,9 @@ t/op/filetest_stack_ok.t		See if file tests leave their argument on the stack
 t/op/for.t				See if for loops work
+t/op/for-control-scope.t		See if scope of for-control control variable(s) works
 t/op/for-many.t				See if n-at-a-time for loops work
+t/op/for-over-scope.t			See if scope of for-over cursor variable(s) works
```
**Commit [tests] New tests focused on scope behaviour of loop variables**

The `for (;;)` variant is now called "for-control" and the `for $var ()` variant is called "for-over" and/or "for-over cursor".
Did you come up with these names or are they based on something?

happy-barney replied:
They are based on loop behaviour:
- `for-control` - for loop with control variable (C-like syntax)
- `for-over` - loop over list

bram-perl replied:
Preamble: when I'm looking at code/when reviewing I tend to be picky about names because I want names to be expressive and instantly make it clear what it is.

Let me clarify the question: did you come up with these names or are they already well established somewhere?

The reason why I'm asking: before this PR I would have no clue what a "for control" loop would mean (I - obviously - know the construct but I've never seen it being referred to as a "for control" loop. perlsyn for example refers to it as c-style loop but that also is not a great name). Same with "for over": **I** would not immediately link it with `for (@list)`

**If** the names are already established somewhere then it's my knowledge that is lacking and then the names are good.
**If** you came up with the names then it might be worthwhile to think about/discuss alternative/better names so that it is immediately clear to everyone what it means/what it is used for. 

Just an initial thought `for-list` might be clearer then `for-over`.
Or even `for-over-list` would be better then `for-over`.

happy-barney replied:
ok, now I understand your question.

No, these exact names were not used in code base so far.
I though about `list` variant as well but it is not always over list. More likely it is `for-over-iterable-with-cursor` (in future can be object "implementing interface iterable" or result of function containing "yield")

I also though about introducing new directory (eg: `grammar/expression-for-control-scope.t`) where tests will be named after respective grammar rule they are testing (in this case with suffix what is tested).

Regarding discussion: IMHO PR review is proper place. Rest kind of discussions tends to stale.

Few other ideas (used also LLM to suggest some as well):
`for-control`:
- `for-with-control`
- `for-traditional`
- `for-with-condition`
- `for-threepart`

`for-over`:
- `for-with-cursor`
- `for-iterate`
- `for-iterator`
- `for-values`
- `for-in`

I wanted to avoid names like:
- `for-each` - there are keywords `foreach` and `each` which may derail though process
- `for-list` - word `list` represents special structure in language, which differs from `array` and `hash`

bram-perl replied:
I've still been thinking about better names for this..

**For `for-control`:**
I would be tempted to call it something like `c-style-for-loop` or `for-c-style` or something in that style.
The term "C-style for loop" appears to be well established (it's also used in our docs) and I think most will instantly recognize it.

**For `for-over`:**
I still haven't come up with a good name..

Going over your names:
- `for-with-cursor`: I dislike the `cursor` part. 
- `for-iterate`: no opinion
- `for-iterator`: no opinion
- `for-values`: not convinced, could be confused for something like: `for (values %xxx) {}`
- `for-in`: also not convinced

Checking with Gemini:
- `Range-based for loop`: appears to be official name in C++ (since C++11)
- `For-Each Loop`: but I agree that this one can be confusing. (Especially since for/foreach are aliases)
- `Enhanced For Loop`: used as official term in Java, but I don't like it very much. "Enhanced" is pretty meaningless.

Thinking a bit about `range-based-for` as name: I'm not entirely convinced.. the "range" part to me does not immediately make me think of a list/array.

I think the options I like most (or dislike the least) are: `for-iterator` and `range-based-for`

happy-barney replied:
As I had more time to think about:

`for-control`
- as part of longer rule `bare_statement_for_control`
- stands for "for loop with control expression"
- using `for-cstyle` refers to "as implemented in C", so when C does something, like C++ did, it may be again confusing

`for-over`
- as part of longer rule `bare_statement_for_over`
- stands for "for loop over collection"
- `for-iterate` may be accurate as well, with meaning "cursor iterates over iterable collection"
  - with which leads to longer description: "for loop with cursor iterating over collection"
  - `over` is in both variant, I originally chose shorter version
- - I'm also looking forward when there will be generators available in Perl, where (as non-English speaker) I don't feel that `iterate` is proper word to describe "over generator"
- one current objection I have (and putting that one point difference for my personal winner) is usage of `for-control` version in form (whether variable name is `i`, `c`, `x`, `iter`, ...)
```
for (my $interator = $first; $iterator; $iterator = $iterator->next) {
}
```

So I'd stay with `for-control` and, if after reading my objections you are still more comfortable with `for-iterate` I'd rename `for-over`.

Anyway, it also looks like I should adapt documentation as well :-( maybe even delta :-(
Plus question "where to put such kind of tests?" I asked on mailing list (and already 4 days without single response.

bram-perl replied:
> `for-control`
>
> * using `for-cstyle` refers to "as implemented in C", so when C does something, like C++ did, it may be again confusin

I guess that depends on context. To me a c-style for loop doesn't say anything how it's implemented.

But you will never convince me that `for-control` is the right name: the name is not obvious to what it express. \
I think it's time for other to speak up what they think about the name and if they would instantly know what `for-control` parses.

happy-barney replied:
ok, posting to mailing list

happy-barney replied:
using `for-cstyle`

leonerd commented on MANIFEST:
```diff
@@ -6351,7 +6351,9 @@ t/op/filetest_stack_ok.t		See if file tests leave their argument on the stack
 t/op/filetest_t.t			See if -t file test works
 t/op/flip.t				See if range operator works
 t/op/for.t				See if for loops work
+t/op/for-control-scope.t		See if scope of for-control control variable(s) works
```
"for-cstyle"

happy-barney replied:
done

leonerd commented on perly.y:
```diff
@@ -345,9 +349,14 @@ bare_statement_field_declaration
+	|	bare_statement_for_over
+	;
+
+bare_statement_for_control
```
`bare_statement_for_cstyle`

happy-barney replied:
done

leonerd commented on perly.y:
```diff
@@ -376,106 +385,22 @@ bare_statement_for
-		PERLY_PAREN_OPEN
+	;
+
+bare_statement_for_over
```
`bare_statement_for_itervars`

happy-barney replied:
done

leonerd commented on perly.y:
```diff
@@ -761,6 +678,55 @@ bare_statement_yadayada
 		}
 	;
 
+clause_for_over_cursor
```
`for_itervars`

happy-barney replied:
done

leonerd commented on perly.y:
```diff
@@ -761,6 +678,55 @@ bare_statement_yadayada
+		}
+	;
+
+clause_mexpr
```
Hard to think of a good name here, but maybe `mexpr_in_parens`?

happy-barney replied:
Let's let this one as it is. At the moment it seems reasonable (at least to me).

I'm introducing `clause_*` rules (more in queue) as non-terminals providing mapping of `*` behaviour to reduce number of branches or unnecessary noise.

leonerd commented on MANIFEST:
```diff
@@ -6351,7 +6351,9 @@ t/op/filetest_stack_ok.t		See if file tests leave their argument on the stack
 t/op/for.t				See if for loops work
+t/op/for-control-scope.t		See if scope of for-control control variable(s) works
 t/op/for-many.t				See if n-at-a-time for loops work
+t/op/for-over-scope.t			See if scope of for-over cursor variable(s) works
```
`for-itervar-scope`

happy-barney replied:
done

leonerd reviewed (approved):
Added some suggestions about

Commit:
[parser] Formalize KW_FOR as a keyword starting new lexical scope

Tests covering this changes:
- t/op/for-control-scope.t
- t/op/for-over-scope.t

* Motivation

Normalize where new lexical context starts so every variant
of for loop will start it at same position.

Such normalization will later allow reduction of code currently
duplicated.

* Change description

For statement already defined its own lexical scope in each its branch
though in different places for each branch (using <*> as a marker):

```
for (<*> $control = 0; ...)
for my <*> $cursor (...)
for my <*> ($cursor_a, $cursor_b) (...)
for $cursor (<*> ...)
for my \ <*> $cursor (...)
for \ my <*> $cursor (...)
for \ $cursor (<*> ... )
for (<*> ...)
```

This change formalizes every branch to start with:
```
for <*> ...
```

* Technical details

Lexical scope is started by non-terminal `remember` - empty with an action.

Bison's LALR(1) grammar evaluates action before looking which terminal token
follows, which means, that when `remember` occurs it must be either:
- preceded by unique sequence of terminal tokens (approach used before)
- be on same position for every possible branch (approach used by this change)

As far as `KW_FOR` (as statement) always starts new lexical scope,
it's possible to formalize it by starting it right after keyword,
when `for` statement is identified.

Commit:
[parser] Formalize KW_FOR as a keyword starting new lexical scope

Tests covering this changes:
- t/op/for-control-scope.t
- t/op/for-over-scope.t

* Motivation

Normalize where new lexical context starts so every variant
of for loop will start it at same position.

Such normalization will later allow reduction of code currently
duplicated.

* Change description

For statement already defined its own lexical scope in each its branch
though in different places for each branch (using <*> as a marker):

```
for (<*> $control = 0; ...)
for my <*> $cursor (...)
for my <*> ($cursor_a, $cursor_b) (...)
for $cursor (<*> ...)
for my \ <*> $cursor (...)
for \ my <*> $cursor (...)
for \ $cursor (<*> ... )
for (<*> ...)
```

This change formalizes every branch to start with:
```
for <*> ...
```

* Technical details

Lexical scope is started by non-terminal `remember` - empty with an action.

Bison's LALR(1) grammar evaluates action before looking which terminal token
follows, which means, that when `remember` occurs it must be either:
- preceded by unique sequence of terminal tokens (approach used before)
- be on same position for every possible branch (approach used by this change)

As far as `KW_FOR` (as statement) always starts new lexical scope,
it's possible to formalize it by starting it right after keyword,
when `for` statement is identified.

Commit:
[parser] Extract for-itervars itervars declarations into `clause_itervars`

Extract multiple ways how to declare itervars variable(s) of for-itervars
loop into dedicated non-terminal allowing to merge `for-itervars` branches.

Merged for-over branches of:
- `for $scalar (...)`
- `for my $scalar (...)`
- `for my ($foo, $bar) (...)`
- `for \ $scalar (...)`
- `for my \ $scalar (..)`

Branch `for ()` must still remain as far as LALR(1) grammar will run
into shift/reduce conflict with `for (;;)` - there will be two possible
paths between `KW_FOR` and `(`:
- `KW_FOR remember '('` - from `for (;;)` branch
- `KW_FOR remember clause_itervars '(` - from `for ()` (using default itervars)

Commit:
[parser] Extract for-itervars itervars declarations into `clause_itervars`

Extract multiple ways how to declare itervars variable(s) of for-itervars
loop into dedicated non-terminal allowing to merge `for-itervars` branches.

Merged for-over branches of:
- `for $scalar (...)`
- `for my $scalar (...)`
- `for my ($foo, $bar) (...)`
- `for \ $scalar (...)`
- `for my \ $scalar (..)`

Branch `for ()` must still remain as far as LALR(1) grammar will run
into shift/reduce conflict with `for (;;)` - there will be two possible
paths between `KW_FOR` and `(`:
- `KW_FOR remember '('` - from `for (;;)` branch
- `KW_FOR remember clause_itervars '(` - from `for ()` (using default itervars)

jkeenan commented:
Well, I do not at all like the idea of modifying our test infrastructure and then using those modifications in the very same pull request.  I realize that may not be a popular opinion, but that's the way I see it.  Thumbs down for me on this pull request.
